### PR TITLE
Refactor Code and add StaffBotRequest

### PR DESCRIPTION
**Pull request by kkamalov**

### CODE REVIEW

File: orchestra/models/communication/models.py [hunk 1]
```diff
@@ -73,6 +73,25 @@ class Meta:
         unique_together = (('worker', 'communication_type'),)
 
 
+class StaffBotRequest(StaffBotRequestMixin, BaseModel):
```
Comment (joshblum) [inline orchestra/models/communication/models.py:78]: add docstring like other models

File: orchestra/utils/task_lifecycle.py [hunk 2]
```diff
@@ -133,6 +133,26 @@ def is_worker_certified_for_task(
...
+    if role_counter == 0:
+        return WorkerCertification.Role.ENTRY_LEVEL
+    return WorkerCertification.Role.REVIEWER
+
+
+def role_counter_required_for_new_task(task):
+    if task.status not in [Task.Status.AWAITING_PROCESSING,
+                           Task.Status.PENDING_REVIEW]:
+        raise TaskAssignmentError('Status incompatible with new assignment')
+    assignments_count = task.assignments.count()
```
Comment (joshblum) [inline orchestra/utils/task_lifecycle.py:151]: why not:

`return task.assignments.fitler(assignment_counter__lte=assignments_count).order_by('assignment_counter').first().assignment_counter`

Comment (joshblum) [inline orchestra/utils/task_lifecycle.py:151]: and we should test this method it's funcky

File: orchestra/models/communication/models.py [hunk 3]
```diff
@@ -89,29 +108,14 @@ class StaffingRequestInquiry(StaffingRequestInquiryMixin, BaseModel):
             Description of the project
```
Comment (joshblum) [inline orchestra/models/communication/models.py]: update doc string to remove the fields no longer there

File: orchestra/migrations/0061_auto_20160505_2126.py [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+                ('project_description', models.TextField(blank=True, null=True)),
+                ('status', models.IntegerField(choices=[(0, 'processing'), (1, 'complete')], default=0)),
+                ('task', models.ForeignKey(on_delete=django.db.models.deletion.CASCADE, related_name='staffing_requests', to='orchestra.Task')),
+            ],
+            options={
+                'abstract': False,
+            },
+            bases=(orchestra.models.communication.model_mixins.StaffBotRequestMixin, orchestra.utils.models.DeleteMixin, models.Model),
+        ),
+        migrations.RemoveField(
```
Comment (joshblum) [inline orchestra/migrations/0061_auto_20160505_2126.py]: all of these need `# manually-reviewed` (landing a bug that was preventing this check now)

Comment (kkamalov) [inline orchestra/migrations/0061_auto_20160505_2126.py]: cool

File: orchestra/models/communication/models.py (hunk 1 above)
Comment (kkamalov) [inline orchestra/models/communication/models.py:78]: done

File: orchestra/models/communication/models.py (hunk 3 above)
Comment (kkamalov) [inline orchestra/models/communication/models.py]: done

File: orchestra/utils/task_lifecycle.py [hunk 5]
```diff
@@ -140,9 +140,17 @@ def get_role_from_counter(role_counter):
...
+    to be assigned for task.
+    """
     if task.status not in [Task.Status.AWAITING_PROCESSING,
                            Task.Status.PENDING_REVIEW]:
         raise TaskAssignmentError('Status incompatible with new assignment')
+    elif (task.status == Task.Status.AWAITING_PROCESSING and
+          (task.assignments.filter(assignment_counter=0)
+           .exclude(status=TaskAssignment.Status.FAILED).exists())):
+        raise TaskAssignmentError('Task is in incorrect state')
     assignments_count = task.assignments.count()
```
Comment (joshblum) [inline orchestra/utils/task_lifecycle.py:151]: I think you should update this to the query i suggested in the previous comments so we don't have to hit the DB so many times.

File: orchestra/utils/task_lifecycle.py [hunk 6]
```diff
@@ -171,6 +179,11 @@ def remove_worker_from_task(worker_username, task_id):
     task = Task.objects.get(id=task_id)
     task_assignment = TaskAssignment.objects.get(worker=worker, task=task)
 
+    # mark currently active task assignment as submitted
```
Comment (joshblum) [inline orchestra/utils/task_lifecycle.py]: why do we do this?

File: orchestra/utils/task_lifecycle.py [hunk 7]
```diff
@@ -290,6 +303,14 @@ def reassign_assignment(worker_id, assignment_id):
     if assignment.task.is_worker_assigned(worker):
         raise TaskAssignmentError('Worker already assigned to this task.')
 
+    if assignment.status == TaskAssignment.Status.FAILED:
```
Comment (joshblum) [inline orchestra/utils/task_lifecycle.py]: This state is only for machine tasks actually, we need to refactor the restaffing logic to remove setting a human task to the `FAILED` state: https://github.com/unlimitedlabs/orchestra/blob/master/orchestra/utils/task_lifecycle.py#L158

Comment (joshblum) [inline orchestra/utils/task_lifecycle.py]: don't we want to remove this?

File: orchestra/cron.py [hunk 8]
```diff
@@ -0,0 +1,7 @@
+import kronos
+from orchestra.communication.staffing import send_staffing_requests
+
+
+@kronos.register('*/1 * * * *')
```
Comment (joshblum) [inline orchestra/cron.py]: `@kronos.register('* * * * *')  # run every minute`

**Comment by joshblum** [comment]
lgtm after cron fix

**Pull request merged**